docs/proposals/update-machine.md

```diff
@@ -0,0 +1,163 @@
+ 
+
+```mermaid
+%%{ init: {
+    'themeVariables':
+        { 'fontSize': '12px'}
+} }%%
+flowchart TD
+
+
+end1(("end"))
+begin((" "))
+medretry["return MediumRetry, err"]
+shortretry["return ShortRetry, err"]
+medretry-->end1
+shortretry-->end1
+
+begin-->AddBootstrapTokenToUserData
+-->gms["statusResp,statusErr=driver.GetMachineStatus(...)"]
+-->chkstatuserr{"Checkk statusErr"}
+chkstatuserr--notFound-->chknodelbl{"Chk Node Label"}
+chkstatuserr--else-->createFailed["lastop.Type=Create,lastop.state=Failed,phase=CrashLoopBackOff|Failed"]-->medretry
+chkstatuserr--nil-->initnodename["nodeName = statusResp.NodeName"]-->setnodename
+
+
+chknodelbl--notset-->createmachine["createResp, createErr=driver.CreateMachine(...)"]-->chkCreateErr{"Check createErr"}
+
+chkCreateErr--notnil-->createFailed
+
+chkCreateErr--nil-->getnodename["nodeName = createResp.NodeName"]
+-->chkstalenode{"nodeName != machine.Name\n//chk stale node"}
+chkstalenode--false-->setnodename["if unset machine.Labels['node']= nodeName"]
+-->machinepending["if empty/crashloopbackoff lastop.type=Create,lastop.State=Processing,phase=Pending"]
+-->shortretry
+
+chkstalenode--true-->delmachine["driver.DeleteMachine(...)"]
+-->permafail["lastop.type=Create,lastop.state=Failed,Phase=Failed"]
+-->shortretry
+
+subgraph noteA [" "]
+    permafail -.- note1(["VM was referring to stale node obj"])
+end
+style noteA opacity:0
+
+
+subgraph noteB [" "]
+    createFailed-.- note2(["Proposal: Enhance handling to support startup logic"])
+end
+```
+
+### Enhancement of MCM triggerCreationFlow
+
+The following changes are proposed with a view towards minimal impact on current code and no introduction of a new Machine Phase.
+
+1. We need to introduce enhancements around the call to `Driver.CreateMachine` in the flow illustrated above.
+1. Observe that after the call to a successful `Driver.CreateMachine`, the machine phase is set to `Pending`, the `LastOperation.Type` is currently set to `Create` and the `LastOperation.State` set to `Processing`before returning with a `ShortRetry`. The `LastOperation.Description` is (unfortunately) set to the fixed message: `Creating machine on cloud provider`.
+1. Observe that after an erroneous call to `Driver.CreateMachine`, the machine phase is set to `CrashLoopBackOff` or `Failed`  (in case of creation timeout).
+1. We propose introducing a new MC error code `codes.Startup` indicating that the VM Instance was created but there was a startup error in post VM creation activities. The implementor of `Driver.CreateMachine` can return this error code, indicating that `CreateMachine` needs to be called again. However, the Machine Controller will keep the machine phase for such a situation as  `Pending` and not change the phase to `CrashLoopBackOff` when encountering a `codes.Startup` error. (The instance for the machine HAS been created hence `CrashLoopBackoff` is inappropriate)
```

Comment:
If we decide to use another error code then do change it here as well. Startup signals that the VM was created and the startup of the VM which involves running the user config script, starting up kubelet and other k8s components failed. But here we are hinting at an update call to the VM post create which does not signal `Startup`.

Reply:
One can follow simple terminology - PreCreate, Create, PostCreate.

Reply:
Calling this `InstanceInitialization`